xdsl/dialects/ptr.py

```diff
@@ -0,0 +1,104 @@
+########
+#  This is a port of the 'ptr' dialect proposed in this thread https://discourse.llvm.org/t/rfc-ptr-dialect-modularizing-ptr-ops-in-the-llvm-dialect/75142/21
+#  Main parts of the dialect has already been agreed upon but only the basic datatype is implemented upstream.
+#  When the upstream implementation is merged we should fix our implementation to be consistent with the mlir version.
+#  Current diviations:
+#  (1) Name of the dialect should be changed: ptr_xdsl -> ptr. We currently chose another name to avoid conflicts when loading our dialect to mlir-opt.
+#  (2) There is no ptr.to_ptr operation currently proposed, but there is a memref.to_ptr. We do this so that we can feed the results of convert-memref-to-ptr pass without any conflict.
+########
+
+
+from xdsl.dialects.builtin import IntegerAttrTypeConstr, MemRefType, UnitAttr
+from xdsl.ir import Dialect, ParametrizedAttribute, TypeAttribute
+from xdsl.irdl import (
+    Attribute,
+    IRDLOperation,
+    base,
+    irdl_attr_definition,
+    irdl_op_definition,
+    operand_def,
+    opt_prop_def,
+    prop_def,
+    result_def,
+)
+
+
+@irdl_attr_definition
+class PtrType(ParametrizedAttribute, TypeAttribute):
+    name = "ptr_xdsl.ptr"
+
+
+@irdl_op_definition
+class PtrAddOp(IRDLOperation):
+    name = "ptr_xdsl.ptradd"
+
+    addr = operand_def(PtrType)
+    offset = operand_def(IntegerAttrTypeConstr)
+    result = result_def(PtrType)
+
+    assembly_format = "$addr `,` $offset attr-dict `:` `(` type($addr) `,` type($offset) `)` `->` type($result)"
+
+
+# haven't managed to pass a type here yet. so did it with a hack.
+@irdl_op_definition
+class TypeOffsetOp(IRDLOperation):
+    name = "ptr_xdsl.type_offset"
+
+    elem_type = prop_def(base(Attribute))
```

Comment:
```suggestion
    elem_type = prop_def(Attribute)
```
You're right, I didn't realise that `prop_def` needed a constraint. Also `AnyAttr()` would work here.